### ISSUE TITLE
Aubio for .NET !

### ISSUE THREAD

**Issue by aybe** [aubio/aubio#140]
Here it is : https://github.com/aybe/aubio.net

It is brand new so I cannot attest that everything will work out of the box, but preliminary testing shown no defects.

It took me about 7 days to get this up, manually ... some could say I should have used code generation, thing is, I've looked at it but SWIG does not support .NET Standard and it wouldn't have solved some of the structural problems, unless spending significant time to master and enhance it; which I guess would have exceed by far the time it took me to write it ! Anyway, it's done now and since aubio API slowly evolves it will be a breeze to update the .NET interface.

While coding it, I found a couple of inconsistencies in aubio, mostly missing getters and so on, I have wrote them down and will open another issue about them.

So now, I will be able to get back to coding my audio app, use aubio for that, and fix things as they come up :)

**Comment by piem** [comment]
hello!

wow, that's great! Thank you for your work @aybe.

To be completely honest, I have never touched a line of CSharp, so I'm not the best to evaluate it, but i'm sure this will be useful.

Do you think there could be a way to reduce the number of lines by using macros? Let us know how it evolve!

best, piem

**Comment by aybe** [comment]
Thanks !

Quite a bit happened since,

I did try with another generator : [CppSharp](https://github.com/mono/CppSharp). While interesting at first, dealing with an AST was in fact not funny ! I've been able to hack some parts of it to generate a better interface from aubio since it couldn't map opaque types you've put in place; it was generating two classes which obviously was unfriendly. Problem was solved but a couple of others surfaced, as more and more hacking was needed and the non existent documentation, I just decided to drop it too.

Btw, there are no macros, this was interesting but in the end would not excel manual crafting as I did with aubio sources on the left and my project on the right. Effectively generation was a breeze with it, but expanding it to a better form was a time-consuming job I felt was not worth the effort.

Now for the second part, yesterday I've finished version 2 of the wrapper (https://github.com/aybe/aubio.net/tree/develop) which I'm happy about and is the definitive one:
 - everything is done: spectral, detection, filters, vecutils etc ...
 - hand written and finely tuned (constructors, LINQ etc ... friendly patterns you are used to in .NET)
 - scrapped the previous PRs I've submitted to you, making it friendly internally like [PitchDetection](https://github.com/aybe/aubio.net/blob/develop/Aubio.NET/Detection/PitchDetection.cs) instead of a string, and as a result effectively using a vanilla build of aubio.
 - it supports all configurations: x86, x64 or AnyCPU
 - using binaries built from vcpkg, they are 32/64-bit, have libav, libsndfile, FLAC, OGG, i.e. everything
 - replicated a couple of samples from aubio.org

[By looking at the code](https://github.com/aybe/aubio.net/tree/develop/Aubio.NET) you can see it's much more clear and anyone familiar with aubio will be in known territory.

Still a few issues regarding the support of UWP platform as they stopped supporting custom marshaling, not a big deal as I'll just have to pass pointers instead, a matter of hours of simple but boring code fixes.

**TL;DR** the new version implements all of aubio and better, it just needs a bit more polishing.

Here I'm testing through NuGet packages I'm generating and was wondering whether you'd be okay to have aubio in [https://www.nuget.org/](https://www.nuget.org/) once it's done ?

When it's finished I'll be able to finally use aubio, my main motivation in trying aubio was to see if it could help out for a DJ-ing app., in detecting tempo, adding cues but also color waveforms. I guess it should be of great help for that, you can see my previous attempts here : https://dsp.stackexchange.com/questions/8997/coloring-a-waveform-with-spectral-centroid-or-by-other-means
